File: CodeChef/LongChallenge/SeptLongChallenge2019B/CHEFINSQ.cpp
//https://www.codechef.com/SEPT19B/problems/CHEFINSQ
//Wrong Answer => (only 1 Test case passing) => Possible reason: Integer Overflow! Try BigInteger in Java
//This is getting accepted in Java except for some corner case (6 out of 7 subtask passing)!

#include<bits/stdc++.h>
#define ll long long int
using namespace std;

ll factorial(ll n)
{
    ll result = 1;

    while(n >= 1)
        result *= n--;

    return result;
}

ll combination(ll n, ll r)
{
    return factorial(n)/(factorial(n-r)*factorial(r));
}

int main()
{
    #ifdef DEBUG
        freopen("input.txt", "r", stdin);
        freopen("output.txt", "w", stdout);
    #endif

    ios_base :: sync_with_stdio(false);
    cin.tie(0);

    int t;

    cin >> t;

    while(t--)
    {
        int n,k,ele;

        cin >> n >> k;

        int freq[101] = {0}; //store count of all A[i]


        for (size_t i = 0; i < n; i++)
            {
                cin >> ele;
                freq[ele]++;
            }

            // for (size_t i = 1; i <= 100; i++)
            // {
            //     //cout << "freq of " << i << " => " << freq[i] << "\n";
            // }
            

        int count = 0;
        ll result = 0;

        for (size_t i = 1; i <= 100; i++)
        {
            if(count + freq[i] > k)
                {
                    //cout << "count + freq[" << i << "] = " << count + freq[i] << "\n";
                    result = combination(freq[i], k-count); // => nCR
                    //cout << "freq[i] = " << freq[i] << " | count-k = " << count-k << " | nCr = " << result << "\n";
                    break;
                }
            else
                {
                    count += freq[i];
                    //cout << "count = " << count << "\n";
                }
                
            
        }

        cout << result << "\n";
        
        
    }

    return 0;
}

File: CodeChef/LongChallenge/SeptLongChallenge2019B/CHEFINSQ.java
//https://www.codechef.com/SEPT19B/problems/CHEFINSQ
//Accepted

/* package codechef; // don't place package name! */

import java.util.*;
import java.lang.*;
import java.io.*;
import java.math.*;

/* Name of the class has to be "Main" only if the class is public. */
class Codechef
{
    
    
    static BigInteger factorial(BigInteger n)
	{
		//System.out.println("To be calculated: factorial of " + n);
		BigInteger res = new BigInteger("1");
		BigInteger one = new BigInteger("1");
		
		//System.out.println("compare value" + n.compareTo(one));
		while(n.compareTo(one) == 1)
			{
				res = res.multiply(n);
				n = n.subtract(one);
				//System.out.println("n = " + n);
				
			}
		//System.out.println("res => " + res);
		return res;
	}
	
	static BigInteger  combination(BigInteger n, BigInteger r) {
		return factorial(n).divide((factorial(n.subtract(r)).multiply(factorial(r))));
	}
	
	public static void main (String[] args) throws java.lang.Exception
	{
		// your code goes here
		
		// TODO Auto-generated method stub
		
		int t;
		
		Scanner sc = new Scanner(System.in);
		
		t = sc.nextInt();
		
		while(t > 0)
		{	
			long n,k;
			int ele;
			
			n = sc.nextLong();
			k = sc.nextLong();
			
			long freq[]= new long[101];
			
			for(int i = 0; i < n; i++)
				{
					ele = sc.nextInt();
					freq[ele]++;
				}
			
			BigInteger res = new BigInteger("0");
			long count = 0;
			

			
            int i = 1;
            while(count < k && i <= 100)
            {
                count += freq[i];
                if(count >= k)
                    break;
                i++;
            }

                res = combination(BigInteger.valueOf(freq[i]), BigInteger.valueOf(count-k)); // => nCR

        
			
			System.out.println(res);
			
			t--;
		}
		
	
	}
}
